design-documents/graph-ql/schema-development.md

```diff
@@ -0,0 +1,108 @@
+
+## GraphQL Server
+
+Within the JavaScript ecosystem, there are [a few server choices available](https://www.npmtrends.com/apollo-server-vs-koa-graphql-vs-prisma), but the most popular and mature choice is [Apollo Server][]. Optionally, [`graphql-yoga`](https://www.npmjs.com/search?q=graphql-yoga) uses [Apollo Server][] under the hood.
+
+## Code-First vs. SDL
+
+As [previously mentioned](#graphql-ecosystem), there are myriad JavaScript libraries available to assist in the creation of a [GraphQL][] server, but before we explore them, we should consider the pros & cons of defining our schema as code-first vs. SDL. Many articles [have been written](#references) about this topic and there are excellent arguments on both sides, but code-first appears to be the direction things are headed.
+
+In this document, we will explore the code-first approach.
+
+> Instead of increasing the complexity of GraphQL server development with more tools, we should strive for a simpler development model. Ideally, one that lets developers leverage the programming language they're already using – [this is the idea of code-first](https://www.prisma.io/blog/the-problems-of-schema-first-graphql-development-x1mn4cb0tyl3?source=post_page-----cf0e50d5ccff----------------------#code-first-a-language-idiomatic-way-for-graphql-server-development).
+
+### Pros
+
+- SDL is a generated artifact, not manually maintained.
+- The code-first schema is the single source of truth.
+   - No inconsistencies. Schema definition & resolvers are always in sync.
+   - No redundancy (duplication).
+   - Easier to maintain.
+- The shape of the API is closer to the data model.
+- Better IDE support & developer experience.
+- Better code reuse and type safety.
+- Can split by features/modules.
+- Terse (less boilerplate).
+- Declarative.
+- Scalable.
+
+### Cons
+
+- Some of [the latest code-first libraries](https://www.prisma.io/blog/introducing-graphql-nexus-code-first-graphql-server-development-ll6s1yy5cxl5) are not yet mature enough.
+- Tightly coupled to a specific language (e.g., [TypeScript][]).
+
+## Code-First Frameworks
+
+Both [TypeGraphQL](#typegraphql) and [Nexus](#nexus) are the key players that provide a code-first solution, but they go about it in entirely different ways. One thing they have in common, however, is that they are designed to use [TypeScript][].
+
+_At this point, we're talking about developing a [GraphQL][] API in [Node.js][] with [TypeScript][]._
+
+### [TypeGraphQL][]
+
+> [TypeGraphQL][] is a library that makes this process enjoyable by defining the schema using only classes and a bit of decorator magic.
+
+```ts
+@ObjectType()
+class Recipe {
+  @Field()
+  title: string;
+
+  @Field(type => [Rate])
```

Comment:
Is there a reason why the type cannot be deducted in this case?

Reply:
I agree that this is a weakness of TypeGraphQL's impementation. This is their example, pulled directly from their site, but I'm positive this can be improved. I mean, it _can_ be, but at what cost? We could intercept the AST at compile time to infer these types. There might even be open issues about this.